### ISSUE TITLE
During extra iteration for analysis: JavaScript heap out of memory

### ISSUE THREAD

**Issue by Isinlor** [nolanlawson/fuite#86]
Hi,

I'm using your awesome tool to find out memory leaks in my project. The whole setup works fine across many different locations in my app. However, it fails on this one with fatal error, somewhat ironically with heap out of memory.

I do have leaks in my app, but they are mostly ~2MB. This one seems to be 4GB and happens on "Extra iteration for analysis" which makes me think that it may be not directly related to my app. The memory usage is pretty much constant even if I do 17 iterations, but it blows up on the extra.

I'm didn't really have much luck with minimal reproduction.

```
npx fuite --scenario=tests/fuite/scenario.mjs --iterations=1 http://localhost:5173/login

URL       : http://localhost:5173/login
Scenario  : tests/fuite/scenario.mjs
Iterations: 1

TEST RESULTS

--------------------

⠧ Gathering tests...[ '/org/OrgA/payroll' ]
⠦ Test 1/1 - Go to /org/OrgA/payroll and back - Extra iteration for analysis...
<--- Last few GCs --->

[598257:0x6a4ec80]    40692 ms: Scavenge 4024.6 (4110.5) -> 4022.0 (4111.8) MB, 6.20 / 0.00 ms  (average mu = 0.913, current mu = 0.700) allocation failure; 
[598257:0x6a4ec80]    40715 ms: Scavenge 4027.1 (4112.8) -> 4024.6 (4131.8) MB, 7.64 / 0.00 ms  (average mu = 0.913, current mu = 0.700) allocation failure; 
[598257:0x6a4ec80]    40994 ms: Mark-Compact 4042.4 (4135.8) -> 4035.9 (4145.6) MB, 223.42 / 0.00 ms  (average mu = 0.821, current mu = 0.523) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0xc9e850 node::Abort() [node]
 2: 0xb720ff  [node]
 3: 0xec1a70 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [node]
 4: 0xec1d57 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [node]
 5: 0x10d3dc5  [node]
 6: 0x10ebc48 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 7: 0x10c1d61 v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 8: 0x10c2ef5 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 9: 0x109f536 v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [node]
10: 0x1091024 v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawWithImmortalMap(int, v8::internal::AllocationType, v8::internal::Map, v8::internal::AllocationAlignment) [node]
11: 0x10938a6 v8::internal::FactoryBase<v8::internal::Factory>::NewRawOneByteString(int, v8::internal::AllocationType) [node]
12: 0x12052a5 v8::internal::JsonParser<unsigned short>::MakeString(v8::internal::JsonString const&, v8::internal::Handle<v8::internal::String>) [node]
13: 0x120b2bd  [node]
14: 0x120d229 v8::internal::JsonParser<unsigned short>::ParseJson(v8::internal::Handle<v8::internal::Object>) [node]
15: 0xf537fe v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [node]
16: 0x7f75c3699df6 
Aborted (core dumped)

```

Any idea what it may be?

One thing I've noticed is that output.json from other pages is 20MB in size, mostly due to listing indexes of DOM nodes. Is that expected?

Thanks for help.

**Comment by nolanlawson** [comment]
Interesting. "Extra iteration for analysis" is used to find the source of leaks in collections (Array, Objects, etc.). Perhaps you have a lot of objects/arrays/etc.

Since the out-of-memory is in Node itself, could you try increasing the amount of memory you give to Node?

In other words, run:

```bash
mkdir tmp
cd tmp
npm init --yes
npm i fuite
node --max-old-space-size=8000 ./node_modules/.bin/fuite --scenario=tests/fuite/scenario.mjs --iterations=1 http://localhost:5173/login
```

**Comment by Isinlor** [comment]
Thanks for suggestion, but unfortunately even 22GB is not enough.

`node --max-old-space-size=22000 ./node_modules/.bin/fuite --scenario=tests/fuite/scenario.mjs --iterations=1 http://localhost:5173/login`

```
⠋ Test 1/1 - Go to /org/OrgA/payroll and back - Extra iteration for analysis...
<--- Last few GCs --->

[876184:0x5deb870]   157068 ms: Scavenge 21644.2 (22027.1) -> 21641.5 (22046.1) MB, 25.7 / 0.0 ms  (average mu = 0.945, current mu = 0.750) allocation failure; 
[876184:0x5deb870]   159602 ms: Mark-sweep 21650.3 (22047.8) -> 21645.4 (22052.6) MB, 2494.9 / 0.0 ms  (average mu = 0.783, current mu = 0.115) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb95b60 node::Abort() [node]
 2: 0xa9a7f8  [node]
 3: 0xd6f2f0 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xd6f697 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xf4cba5  [node]
 6: 0xf5f08d v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 7: 0xf3978e v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 8: 0xf3ab57 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 9: 0xf1b0a0 v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [node]
10: 0xf12b14 v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawWithImmortalMap(int, v8::internal::AllocationType, v8::internal::Map, v8::internal::AllocationAlignment) [node]
11: 0xf14dc8 v8::internal::FactoryBase<v8::internal::Factory>::NewRawOneByteString(int, v8::internal::AllocationType) [node]
12: 0x1055119 v8::internal::JsonParser<unsigned short>::MakeString(v8::internal::JsonString const&, v8::internal::Handle<v8::internal::String>) [node]
13: 0x1056e36 v8::internal::JsonParser<unsigned short>::ParseJsonValue() [node]
14: 0x105791f v8::internal::JsonParser<unsigned short>::ParseJson() [node]
15: 0xdf4ed3 v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [node]
16: 0x170dfb9  [node]
Aborted (core dumped)
```

**Comment by nolanlawson** [comment]
That's interesting, but unfortunately, without a reproducible test, I'm not sure how to resolve this. I don't even know what is taking up the extra memory exactly...

**Comment by nolanlawson** [comment]
If you just want to run `fuite` without the extra iteration, you can edit `./node_modules/fuite/src/index.js and comment out these lines:

https://github.com/nolanlawson/fuite/blob/a004e33efe5a0f5e1d36d3bc2fa49c03990ab52a/src/index.js#L167-L178

**Comment by nolanlawson** [comment]
OK nevermind about the repro, I found a minimal repro. I can get it to crash if I have lots of arrays that leak: a3dfcde

**Comment by nolanlawson** [comment]
Fixed in v4.0.2

**Comment by Isinlor** [comment]
Awesome! Thanks for help, that seems to be solving the problem :) .
